src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs

```diff
@@ -880,41 +881,61 @@ public static TimeZoneInfo FromSerializedString(string source)
         /// <see cref="DisplayName"/>.
         /// This method does *not* throw TimeZoneNotFoundException or InvalidTimeZoneException.
         /// </summary>
-        public static ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones()
+        public static ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones() => GetSystemTimeZones(skipSorting: false);
+
+        /// <summary>
+        /// Returns a <see cref="ReadOnlyCollection{TimeZoneInfo}"/> containing all valid TimeZone's from the local machine.
+        /// This method does *not* throw TimeZoneNotFoundException or InvalidTimeZoneException.
+        /// </summary>
+        /// <param name="skipSorting">If true, The collection returned may not necessarily be sorted.</param>
+        /// <remarks>By setting the skipSorting parameter to true, the method will attempt to avoid sorting the returned collection.
+        /// This option can be beneficial when the method user does not require a sorted list and aims to enhance the performance. </remarks>
+        public static ReadOnlyCollection<TimeZoneInfo> GetSystemTimeZones(bool skipSorting)
         {
             CachedData cachedData = s_cachedData;
 
             lock (cachedData)
             {
-                if (cachedData._readOnlySystemTimeZones == null)
+                if ((cachedData._readOnlySystemTimeZones == null && !skipSorting) || (cachedData._readOnlyUnsortedSystemTimeZones == null && skipSorting))
                 {
-                    PopulateAllSystemTimeZones(cachedData);
-                    cachedData._allSystemTimeZonesRead = true;
+                    if (!cachedData._allSystemTimeZonesRead)
+                    {
+                        PopulateAllSystemTimeZones(cachedData);
+                        cachedData._allSystemTimeZonesRead = true;
+                    }
 
                     if (cachedData._systemTimeZones != null)
                     {
                         // return a collection of the cached system time zones
                         TimeZoneInfo[] array = new TimeZoneInfo[cachedData._systemTimeZones.Count];
                         cachedData._systemTimeZones.Values.CopyTo(array, 0);
```

Comment:
If skipSorting is true and _readOnlySystemTimeZones is not null, we'll hit the else path below that will ignore array, in which case this will have been wasted work/allocation.

Reply:
Instead maybe we can early return `cachedData._readOnlySystemTimeZones` if `cachedData._readOnlySystemTimeZones != null && skipSorting` right after the lock? This would render `cachedData._readOnlyUnsortedSystemTimeZones` as "permanently" `null`, but at that point, we wouldn't need to use more memory to store a copy of the same collection?

Reply:
I addressed that by always initializing `_readOnlyUnsortedSystemTimeZones` when `readOnlySystemTimeZones` get initialized.  This will gaurantee allocating the array only when really needed.